src/graph.jl

```diff
@@ -15,13 +15,92 @@ struct BaseGraph<:AbstractGraph
 end
 
 
+function polygon_array(coords::Vector{Vector{Vector{Vector{Float64}}}})::Array{LibGEOS.Polygon}
+    # Takes a complex vector object and returns an array of Polygons
+    # TODO(matthew): implement this, and add return type to signature
+    return []
+end
+
+
+function min_bounding_rect(coords::Vector{Vector{Vector{Vector{Float64}}}})::Tuple
+    # Takes a complex vector object and returns an (axis-aligned) minimum
+    # bounding rectangle
+    # TODO(matthew): implement this, and add return type to signature
+    return ()
+end
+
+
+function read_table(filepath::AbstractString)::Shapefile.Table
+    """ Read table from shapefile. If a .shp and a .dbf file of the same name
+        are not found, then we throw an error.
+    """
+    prefix = splitext(filepath)[1]
+    if !isfile(prefix * ".shp") || !isfile(prefix * ".dbf")
+        throw(ArgumentError("To read a graph from a shapefile, we require a .shp and .dbf file of the same name in the same folder."))
+    end
+    return Shapefile.Table(prefix)
+end
+
+
+function all_node_properties(table::Shapefile.Table)::Array{Dict{String, Any}}
+    """ Returns an Array of Dictionaries. Each dictionary corresponds to one
+        node in the graph.
+    """
+    properties = propertynames(table) # returns array of symbols
+    string_keys = String.(properties) # convert by broadcasting
+
+    # internal function because we want to use both properties and values
+    function get_node_properties(row::Shapefile.Row)
+        values = map(p -> getproperty(row, p), properties)
+        return Dict(string_keys .=> values)
+    end
+
+    return get_node_properties.(table)
+end
+
+
+function get_node_coordinates(row::Shapefile.Row)::Vector{Vector{Vector{Vector{Float64}}}}
+    """ Construct an array of LibGEOS.Polygons from the given coordinates. The
+        coordinates are structured in the following way. Each element in the
+        outermost array represents one polygon. (One node can be made up of
+        multiple polygons).
+        [
+            [                       # one array = one polygon
+                [                   # points corresponding to outer ring
+                    [1.0, 2.0],     # single x,y coordinate of a point
+                    ...
+                ],
+                [                   # points corresponding to a hole in polygon
+                    [1.5, 1.7],
+                    ...
+                ],
+                ...                 # any other subsequent arrays would
+                                    # correspond to other holes
+            ],
+            ...                     #  subsequent arrays correspond to other
+                                    #  polygons
+        ]
+    """
+    return LibGEOS.GeoInterface.coordinates(getfield(row, :geometry))
+end
+
+
 function graph_from_shp(filepath::AbstractString,
```

Comment:
All that we are doing here is loading in a Shapefile and then calling some functions to extract the "shapes" of the blocks/precincts/counties/whatever is in the Shapefile.